### PR TITLE
Preserve data

### PR DESCRIPTION
**Pull request by Mahamadou-Nouridine**
### Preserve data
In this Pr, my coding partner and I brought changes that include:

- Store and load data in files.
- On exit from the app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. `rentals.json`)
- On startup of the app load all data from files. Remember to handle the case when one or more of the files are missing.

### CODE REVIEW

File: classes/rental.rb [hunk 1]
```diff
@@ -8,26 +8,57 @@ class Rental
...
+      'book' => {
+        'title' => @book.title,
+        'author' => @book.author
+      },
+      'person' => {
+        'type' => @person.class.name,
+        'id' => @person.id,
+        'age' => @person.age,
+        'name' => @person.name,
+        'parent_permission' => @person.parent_permission
```
Comment (Munsa1) [inline classes/rental.rb]: - [ ] Nice work, however, we have an error when the user tries to exit the app.
The error is raised on line 28 of the `rental.rb` file, where the code attempts to call parent_permission on an instance of the Teacher class. Based on the error message, the issue seems to be related to the serialization of the Teacher object into JSON format. The `to_json` method is called on a Teacher object, but the `parent_permission` method is not recognized, leading to the `NoMethodError`.
To fix this issue, you need to ensure that the Teacher class has a parent_permission method defined.

Screenshot of error
image

File: classes/rental.rb [hunk 2]
```diff
@@ -8,26 +8,57 @@ class Rental
   def initialize(date, book, person)
     @date = date
     @book = book
+    book.rentals << self
     @person = person
+    person.rentals << self
```
Comment (Munsa1) [inline classes/rental.rb:12]: - [x] Again there is an error in creating and saving a rental.  It seems like you are trying to add the current rental (self) to the rentals collection of a Book object (book). However, the book object is nil, which means it has not been instantiated properly.
To fix this issue, ensure that you properly create and initialize the book object before using it. You should make sure that you have an instance of the Book class and that it is correctly associated with the Rental class.
 
e.g

```
class Rental
  attr_accessor :date, :book, :person

  def initialize(date, book, person)
    @date = date
    @book = book
    @person = person

    # Make sure to initialize the rentals array for the associated book.
    @book.rentals ||= []
    @book.rentals << self
  end
end
```


Screenshot
image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @Mahamadou-Nouridine @Kel-nana , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Able To Save People And Books :heavy_check_mark: 
Correct Git Flow :heavy_check_mark: 
Descriptive ReadMe File :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
n/a


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by tresorsawasawa — APPROVED** [review]
# STATUS: APPROVE :heavy_check_mark: :heavy_check_mark:  :champagne: 

Hi @Mahamadou-Nouridine and @Kel-nana 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 :clap: :clap: :100: 

nodding-yes

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- Your README file is professional :heavy_check_mark: 

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**